In [2]:
def convert_rgb_hsl(r,g,b):
    R = r / 255
    G = g / 255
    B = b / 255
    minimum = min([R,G,B])
    maximum = max([R,G,B])
    l = ((minimum+maximum)/2)*100
    if l>50:
        s = ((maximum-minimum)/(2.0-maximum-minimum))*100
    else:
        s = ((maximum-minimum)/(maximum+minimum))*100

    if maximum == R:
        h = ((G-B)/(maximum-minimum))*60
    elif maximum == G:
        h = (2.0 + ((B-R)/(maximum-minimum)))*60
    else:
        h = (4.0 + ((R-G)/(maximum-minimum)))*60
    return (h,s,l)

In [98]:
def Read_Image(filename):
    im = Image.open(filename)
    pixel_values = list(im.getdata())
    #im.show()
    #print(pixel_values)
    dict_colors = {}
    for pixel in pixel_values:
        if pixel not in dict_colors:
            dict_colors[pixel] = 1
        else:
            dict_colors[pixel] += 1
    blue_red = [0,0] #store number of blues (0) and reds (1)
    for key in dict_colors.keys():
        if  dict_colors[key] > 20 and key != (255,255,255,255) and key != (0,0,0,255)and key[0]!=key[2]:
            (h,s,l) = convert_rgb_hsl(key[0], key[1], key[2])
            if l> 20:
                if h>(170) and h<(300):#color is blue
                    #print(key)
                    #print('h = '+ str(h))
                    #print('s = '+ str(s))
                    #print('l = '+ str(l))
                    blue_red[0] +=1
                else:#color is red
                    blue_red[1] +=1
    return blue_red
        

In [99]:
def getFeatures(file_list):
    feature_list =[[]]*len(file_list)
    counter = 0
    for file in file_list:
        example = Read_Image(file)
        feature_list[counter]=example
        counter+=1
    return feature_list

In [100]:
def getTrainingLabels(file_list):
    labels = file_list
    counter = 0
    for file in file_list:
        im = Image.open(file)
        im.show()
        labels[counter] = input("Label: ")
        counter += 1
    return labels
        

In [101]:
import os
import numpy as np
from PIL import Image
file_list = os.listdir(os.curdir)
file_list = file_list[2:(len(file_list)-1)]
x = getFeatures(file_list)
print(x)
y = getTrainingLabels(file_list) #optional for labeling data by hand

[[58, 160], [78, 66], [37, 110], [43, 153], [232, 21], [171, 69], [138, 219], [23, 136], [186, 84], [191, 70], [22, 194], [23, 152], [28, 421], [41, 100], [44, 64], [40, 75], [45, 69], [24, 392], [11, 349], [57, 55], [73, 62], [135, 209], [82, 135], [44, 188], [55, 133], [309, 16], [41, 220], [49, 207], [42, 180], [36, 278], [209, 12], [259, 12], [255, 13], [87, 58], [269, 22], [33, 242], [13, 108], [47, 100], [50, 78], [78, 223], [32, 63], [32, 48], [34, 41], [220, 23], [17, 153], [45, 85], [56, 75], [57, 59], [12, 154], [30, 117], [12, 198], [22, 209], [10, 82], [14, 99], [32, 95], [36, 214], [42, 132], [75, 189], [191, 155], [307, 40], [298, 37], [84, 110], [108, 91], [129, 54], [66, 122], [47, 240], [19, 230], [30, 226], [34, 328], [38, 337], [23, 138], [68, 121], [52, 110], [76, 52], [49, 57], [38, 58], [45, 144], [125, 156], [241, 91], [139, 157], [42, 81], [21, 181], [30, 194], [33, 113], [38, 123], [201, 11], [43, 78], [50, 58], [46, 50], [51, 140], [122, 94], [46, 81], [54, 79

In [102]:
print(y)

['okay', 'good', 'okay', 'okay', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'okay', 'good', 'okay', 'bad', 'bad', 'bad', 'good', 'good', 'okay', 'bad', 'good', 'okay', 'okay', 'okay', 'good', 'okay', 'okay', 'okay', 'okay', 'good', 'good', 'good', 'bad', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'bad', 'bad', 'bad', 'good', 'good', 'okay', 'bad', 'bad', 'good', 'okay', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'okay', 'good', 'okay', 'good', 'okay', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'okay', 'okay', 'okay', 'good', 'good', 'okay', 'okay', 'good', 'okay', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'okay', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', '

In [103]:
y = ['okay', 'good', 'okay', 'okay', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'okay', 'good', 'okay', 'bad', 'bad', 'bad', 'good', 'good', 'okay', 'bad', 'good', 'okay', 'okay', 'okay', 'good', 'okay', 'okay', 'okay', 'okay', 'good', 'good', 'good', 'bad', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'bad', 'bad', 'bad', 'good', 'good', 'okay', 'bad', 'bad', 'good', 'okay', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'okay', 'good', 'okay', 'good', 'okay', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'okay', 'okay', 'okay', 'good', 'good', 'okay', 'okay', 'good', 'okay', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'okay', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'okay', 'okay', 'good', 'good', 'bad', 'okay', 'okay', 'good', 'bad', 'okay', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'okay', 'okay', 'bad', 'okay', 'okay', 'bad', 'bad', 'okay', 'good', 'good', 'bad', 'good', 'okay', 'good', 'good', 'okay', 'good', 'okay', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'okay', 'okay', 'okay', 'good', 'good', 'okay', 'good', 'good', 'good', 'good']

In [137]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(x_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [152]:
from sklearn import metrics
p = knn.predict(x_test)
metrics.classification_report(p,y_test)
metrics.accuracy_score(p,y_test)

0.75

|     |  precision   | recall  |f1-score   | support |
|-----|--------------|---------|-----------|---------|
| bad |    1.00      |  0.83   |   0.91    |    6    |
|-----|--------------|---------|-----------|---------|
|good |    0.71      |  0.87   |   0.78    |    23   |
|-----|--------------|---------|-----------|---------|
|okay |    0.73      |  0.53   |   0.62    |    15   |
|-----|--------------|---------|-----------|---------|
|total|    0.76      |  0.75   |   0.74    |    44   |

Accuracy Score = .75

One reason for innaccuracy is that classifications used for training are done bby hand and may be inconsistent.

In [162]:
from sklearn import svm
clf = svm.SVC(gamma=5, C=100000000)
clf.fit(x_train, y_train)  
p2 = clf.predict(x_test)
print(p2)
print(y_test)
metrics.classification_report(p2,y_test)
metrics.accuracy_score(p2, y_test)

['good' 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good'
 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good'
 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good'
 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good' 'good'
 'good' 'good' 'good' 'good']
['bad', 'good', 'okay', 'good', 'good', 'good', 'good', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'okay', 'okay', 'good', 'bad', 'good', 'good', 'good', 'good', 'good', 'okay', 'good', 'okay', 'good', 'okay', 'bad', 'good', 'okay', 'good', 'good', 'okay', 'bad', 'bad', 'good', 'okay', 'good', 'good', 'good', 'good']


/Users/erinstafford/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1115: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


0.63636363636363635

# SVM seems to predict all instances as 'good'

|     |precision    |recall  |f1-score   |support|
|-----|-------------|--------|-----------|-------|
|bad  |    0.00     |0.00    | 0.00      |  0    |
|-----|-------------|--------|-----------|-------|
|good |    1.00     |0.64    | 0.78      |  44   |
|-----|-------------|--------|-----------|-------|
|okay |    0.00     |0.00    | 0.00      |  0    |
|-----|-------------|--------|-----------|-------|
|total|    1.00     |0.64    | 0.78      |  44   |

Accuracy score = 0.63636363636363635


# New get features that takes into account the number of pixels of each color

In [163]:
def Read_Image(filename):
    im = Image.open(filename)
    #im = im.crop((550,100,650,500))
    pixel_values = list(im.getdata())
    #im.show()
    #print(pixel_values)
    dict_colors = {}
    for pixel in pixel_values:
        if pixel not in dict_colors:
            dict_colors[pixel] = 1
        else:
            dict_colors[pixel] += 1
    blue_red = [0,0] #store number of blues (0) and reds (1)
    important_colors = {}
    numpixels = [0,0]
    for key in dict_colors.keys():
        if  dict_colors[key] > 20 and key != (255,255,255,255) and key != (0,0,0,255)and key[0]!=key[2]:
            important_colors[key] = dict_colors[key]
            (h,s,l) = convert_rgb_hsl(key[0], key[1], key[2])
            if l> 20:
                if h>(170) and h<(300):#color is blue
                    blue_red[0] +=1
                    numpixels[0] += dict_colors[key]
                else:#color is red
                    blue_red[1] +=1
                    numpixels[1] += dict_colors[key]
    return blue_red, numpixels
        

In [164]:
def getFeaturesPixels(file_list):
    feature_list =[[]]*len(file_list)
    counter = 0
    for file in file_list:
        [blue_red, num_pixels] = Read_Image(file)
        blue_red.append(num_pixels[0])
        blue_red.append(num_pixels[1])
        feature_list[counter]=blue_red
        counter+=1
    return feature_list

In [170]:
file_list = os.listdir(os.curdir)
file_list = file_list[2:(len(file_list)-1)]
x = getFeaturesPixels(file_list)

# Try KNN

In [230]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(x_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [231]:
from sklearn import metrics
p = knn.predict(x_test)
a=metrics.accuracy_score(p, y_test)
r=metrics.classification_report(p,y_test)
print(a)
print(r)

0.818181818182
             precision    recall  f1-score   support

        bad       1.00      1.00      1.00         3
       good       1.00      0.76      0.86        33
       okay       0.50      1.00      0.67         8

avg / total       0.91      0.82      0.84        44



|      |precision  |  recall | f1-score  | support|
|------|-----------|---------|-----------|--------|
|bad   |   1.00    |  1.00   |  0.75     |   3    |
|------|-----------|---------|-----------|------- |
|good  |   1.00    |  0.76   |  0.86     |   33   |
|------|-----------|---------|-----------|--------|
|okay  |   0.50    |  1.00   |  0.67     |   8    |
|------|-----------|---------|-----------|--------|
|total |   0.91    |  0.82   |  0.84     |   44   |

Accuracy score = 0.818181818182

# Try SVM

In [211]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

In [212]:
from sklearn import svm
clf = svm.SVC(gamma=.007, C=10000)
clf.fit(x_train, y_train)  
p2 = clf.predict(x_test)
a = metrics.accuracy_score(p2, y_test)
r = metrics.classification_report(p,y_test)
print(a)
print(r)

0.522727272727
             precision    recall  f1-score   support

        bad       0.50      0.25      0.33         4
       good       0.61      0.47      0.53        30
       okay       0.21      0.40      0.28        10

avg / total       0.51      0.43      0.45        44



|       |precision  |  recall | f1-score  | support| 
|-------|-----------|---------|-----------|--------|
|bad    |   0.50    |   0.25  |   0.33    |     4  |  
|-------|-----------|---------|-----------|--------|
|good   |   0.61    |   0.47  |   0.53    |    30  | 
|-------|-----------|---------|-----------|--------|
|okay   |   0.21    |   0.40  |   0.28    |    10  |
|-------|-----------|---------|-----------|--------|
|total  |   0.51    |   0.43  |   0.45    |    44  |



Accuracy score = 0.522727272727

# Evaluate your features try a few different learning algorithms. Describe what algorithms and via which tools you have used.  Evaluate various models and report performance of your model on your test data. You can provide the evaluation on your test data or use cross-validation. Please be communicative about it and let me know what is your approach for improving your initial results. Post your main test results on canvas and codes on your GitHub repository.

The algorithms used are K-Nearest Neighbors and Support Vector Machine. These are implemented using the scikit learn python package. There are two different feature sets used. The first feature set is when the number of shades of blue and the number of shades of red are used. The second feature set includes the number of pixels that are  blue and red. For KNN, the second feature set resulted in more accurate classifications with k = 3. For SVM, the first data set had the most accurate classifications. In both cases, KNN was more accurate than SVM. The approach taken to improve the results was to add the features number of pixels that are blue and number of pixels that are red. More data was also uploaded and a code to automatically convert PDF files to a list of png files was implemented. Furthermore, in an attempt to improve the by-hand classification of the train/test data a labeling function was implemented.